src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -276,116 +308,160 @@ starting_address(h::HubbardRealSpace) = h.address
+end
+@inline function _column_components(h::HubbardRealSpace, address::CompositeFS)
+    return _column_components(h, address, address.components, Val(1))
+end
+@inline function _column_components(::HubbardRealSpace, _, ::Tuple{}, ::Val)
+    return ()
+end
+@inline function _column_components(
+    h::HubbardRealSpace, address, (a, as...), ::Val{I}
+) where {I}
+    data = HubbardRealSpaceComponentData{I}(h.geometry, address, a, h.t[I])
+    rest = _column_components(h, address, as, Val(I + 1))
+    return (data, rest...)
 end
 
-# All steps of recursive function (should) get inlined, creating a type-stable tuple of
-# offdiagonals.
-@inline function _get_comp_offdiags((a,as...), h, ::Val{I}) where {I}
-    return (offdiagonals(h, I, a), _get_comp_offdiags(as, h, Val(I+1))...)
+function _split_index_component(column, i)
+    components = column.components
+    chosen_component = 0
+    while i > 0
+        chosen_component += 1
+        i -= index_apply(length, components, chosen_component)
+    end
+    i += index_apply(length, components, chosen_component)
+    return chosen_component, i
 end
-@inline _get_comp_offdiags(::Tuple{}, h, ::Val) = ()
 
-function offdiagonals(h::HubbardRealSpace{C,A}, address::A) where {C,A<:CompositeFS}
-    parts = get_comp_offdiags(h, address)
-    return HubbardRealSpaceOffdiagonals(address, parts, sum(length, parts))
+function random_offdiagonal(column::HubbardRealSpaceColumn)
+    directions = 2 * num_dimensions(column.hamiltonian.geometry)
+    random_number = rand(1:num_offdiagonals(column))
+    component, remainder = _split_index_component(column, random_number)
+
+    addr, val = index_apply(getindex, column.components, component, remainder)
+    return addr, 1/column.num_offdiagonals, val
 end
 
-Base.size(o::HubbardRealSpaceOffdiagonals) = (o.length,)
+struct HubbardRealSpaceColumnOffdiagonals{A,G,C<:Tuple} <: AbstractVector{Pair{A,Float64}}
+    address::A
+    geometry::G
+    components::C
+    num_offdiagonals::Int
+end
 
-# Becomes type unstable without inline for lots of components. Recursive function is used
-# because the type of the result of `o.parts[i]` can not be inferred.
-@inline function Base.getindex(o::HubbardRealSpaceOffdiagonals{A}, chosen) where {A}
-    return _getindex(o.parts, o.address, chosen, Val(1))
+function offdiagonals(column::HubbardRealSpaceColumn)
+    return HubbardRealSpaceColumnOffdiagonals(
+        column.address,
+        column.hamiltonian.geometry,
+        column.components,
+        column.num_offdiagonals,
+    )
 end
-@inline function _getindex((p, ps...), address::A, chosen, comp::Val{I}) where {A,I}
-    if chosen ≤ length(p)
-        new_add, val = p[chosen]
-        return BitStringAddresses.update_component(address, new_add, comp), val
+num_offdiagonals(column) = column.num_offdiagonals
+
+@inline function Base.iterate(ods::HubbardRealSpaceColumnOffdiagonals, state=(1,1,1))
+    i, j, k = state
```

Comment:
What do the components `i`, `j`, and `k` mean? Maybe some code comments (or more descriptive names) would help.